### PR TITLE
Added CSV export for list responses

### PR DESCRIPTION
**Pull request by wachterjohannes**
__tasks:__

- [x] test coverage
- [x] UI
- [x] config dialog
- [x] translations (waiting for FR)
- [x] Snippets (no array)
- [x] Roles
- [x] Categories
- [x] Tags
- [x] Contact
- [x] Account

__informations:__

| q                | a
| ---------------- | ---
| Fixed tickets    | none
| Related PRs      | https://github.com/massiveart/husky/pull/617 https://github.com/sulu-io/sulu-standard/pull/628
| BC breaks        | none
| Documentation PR | none

### CODE REVIEW

File: src/Sulu/Bundle/AdminBundle/Resources/public/js/aura_extensions/sulu-buttons.js [hunk 1]
```diff
@@ -190,6 +190,22 @@
                     }
                 },
                 {
+                    name: 'export',
+                    template: {
+                        title: 'public.export',
+                        icon: 'download',
+                        callback: function() {
+                            var $container = $('<div/>');
+                            $('body').append($container);
```
Comment (wachterjohannes) [inline src/Sulu/Bundle/AdminBundle/Resources/public/js/aura_extensions/sulu-buttons.js]: component stop hisself on dialog closed

File: src/Sulu/Bundle/CoreBundle/DependencyInjection/Compiler/CsvHandlerCompilerPass.php [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    public function process(ContainerBuilder $container)
+    {
+        $id = 'fos_rest.view_handler';
+        if (!$container->hasDefinition($id)) {
+            $id = $container->getAlias($id);
+        }
+
+        $definition = $container->getDefinition($id);
+        $definition->addMethodCall(
+            'registerHandler',
```
Comment (wachterjohannes) [inline src/Sulu/Bundle/CoreBundle/DependencyInjection/Compiler/CsvHandlerCompilerPass.php]: there is no nicer way todo this

File: src/Sulu/Component/Rest/ListBuilder/AbstractListBuilder.php [hunk 3]
```diff
@@ -101,8 +101,7 @@ function (FieldDescriptorInterface $fieldDescriptor) {
                 /** @var PropertyMetadata $propertyMetadata */
                 $propertyMetadata = $fieldDescriptor->getMetadata()->get(PropertyMetadata::class);
 
-                return $propertyMetadata->getDisplay() === PropertyMetadata::DISPLAY_YES
-                    || $propertyMetadata->getDisplay() === PropertyMetadata::DISPLAY_ALWAYS;
+                return $propertyMetadata->getDisplay() !== PropertyMetadata::DISPLAY_NEVER;
```
Comment (wachterjohannes) [inline src/Sulu/Component/Rest/ListBuilder/AbstractListBuilder.php]: should maybe also go to master but there this is not really needed.

File: src/Sulu/Component/Rest/ListBuilder/Metadata/PropertyMetadata.php [hunk 4]
```diff
@@ -58,4 +58,30 @@ public function has($name)
     {
         return array_key_exists($name, $this->metadata);
     }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function serialize()
```
Comment (wachterjohannes) [inline src/Sulu/Component/Rest/ListBuilder/Metadata/PropertyMetadata.php]: already in a hotfix after rebase this should go away

Comment (danrot) [inline src/Sulu/Component/Rest/ListBuilder/Metadata/PropertyMetadata.php]: So can you rebase that? :-)

**Comment by wachterjohannes** [comment]
@danrot this is rebased now

File: src/Sulu/Bundle/CategoryBundle/Resources/public/js/components/categories/list/main.js [hunk 5]
```diff
@@ -37,7 +37,16 @@ define(function () {
                 toolbar: {
                     buttons: {
                         add: {},
-                        deleteSelected: {}
+                        deleteSelected: {},
+                        export: {
+                            options: {
+                                urlParameter: {
+                                    flat: true,
+                                    limit: 999999
```
Comment (danrot) [inline src/Sulu/Bundle/CategoryBundle/Resources/public/js/components/categories/list/main.js]: Shouldn't it be possible to omit this, and really export everything? That's the actual intention of the export, isn't it?

Comment (danrot) [inline src/Sulu/Bundle/CategoryBundle/Resources/public/js/components/categories/list/main.js]: Same question is of course valid for the other configurations like this.

File: src/Sulu/Component/Rest/Csv/ObjectNotSupportedException.php [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+     * @var View
+     */
+    private $view;
+
+    public function __construct(View $view)
+    {
+        parent::__construct('The view is not supported to create a csv-response.');
+
+        $this->view = $view;
+    }
```
Comment (danrot) [inline src/Sulu/Component/Rest/Csv/ObjectNotSupportedException.php]: Would a getter for the view make sense?

File: src/Sulu/Component/Rest/ListBuilder/ListRepresentation.php [hunk 7]
```diff
@@ -67,13 +67,23 @@ class ListRepresentation extends PaginatedRepresentation
...
-     * @param int    $limit      The size of one page
-     * @param int    $total      The total number of elements
+     * @var mixed[]
+     */
+    protected $data;
+
+    /**
+     * @var string
+     */
+    protected $rel;
```
Comment (danrot) [inline src/Sulu/Component/Rest/ListBuilder/ListRepresentation.php]: Why is that needed now? How was that handled before?

Comment (wachterjohannes) [inline src/Sulu/Component/Rest/ListBuilder/ListRepresentation.php]: i use this now for the name of the file before it was only used by the wrapped object (see constructor)

File: src/Sulu/Bundle/CategoryBundle/Resources/public/js/components/categories/list/main.js (hunk 5 above)
Comment (wachterjohannes) [inline src/Sulu/Bundle/CategoryBundle/Resources/public/js/components/categories/list/main.js]: the problem here is, that the rest-endpoint is the same as for the list. therefor we have to fix it there but currently i would add this attribute or should we provide a default value for that?

File: src/Sulu/Component/Rest/Csv/ObjectNotSupportedException.php (hunk 6 above)
Comment (wachterjohannes) [inline src/Sulu/Component/Rest/Csv/ObjectNotSupportedException.php]: fixed

File: src/Sulu/Bundle/CategoryBundle/Resources/public/js/components/categories/list/main.js (hunk 5 above)
Comment (danrot) [inline src/Sulu/Bundle/CategoryBundle/Resources/public/js/components/categories/list/main.js]: This is somehow related to https://github.com/sulu-io/sulu/issues/1011, but on the other hand it's also completely different :smile: 

Actually we said once that there should be always a limit on the REST API, maybe we have to rethink that under this circumstances. Or we can omit this one here, and use null as default limit only for the csv format? /cc @chirimoya

Comment (wachterjohannes) [inline src/Sulu/Bundle/CategoryBundle/Resources/public/js/components/categories/list/main.js]: this have to be handled then in all controllers. i would use maybe a default value (9999999999999) in the csv-export method.

Comment (danrot) [inline src/Sulu/Bundle/CategoryBundle/Resources/public/js/components/categories/list/main.js]: Has it to be handled in all controllers? I don't think so. Could also be handled in the `ListRestHelper::getLimit` method, where it (at least should) apply to all controllers.

Comment (wachterjohannes) [inline src/Sulu/Bundle/CategoryBundle/Resources/public/js/components/categories/list/main.js]: ou yeah thats right i havent thought aber it in deeper. should i change it there to return null for csv format?

Comment (danrot) [inline src/Sulu/Bundle/CategoryBundle/Resources/public/js/components/categories/list/main.js]: In a discussion with @chirimoya we decided that it is safer to implement that in the `getLimit` method only for the CSV format, because it is less a security issue due to the fact that we are using streams for CSV.

File: src/Sulu/Component/Rest/ListBuilder/ListRestHelper.php [hunk 8]
```diff
@@ -76,7 +76,12 @@ public function getSortOrder()
      */
     public function getLimit()
     {
-        return $this->getRequest()->get('limit', 10);
+        $default = 10;
+        if ($this->getRequest()->getRequestFormat() === 'csv') {
+            $default = null;
+        }
+
+        return $this->getRequest()->get('limit', $default);
```
Comment (danrot) [inline src/Sulu/Component/Rest/ListBuilder/ListRestHelper.php]: @wachterjohannes Where is tested if that really works? Are there any test cases with more than 10 entities? Otherwise I would add a unit test here.

Comment (wachterjohannes) [inline src/Sulu/Component/Rest/ListBuilder/ListRestHelper.php]: i have added unit tests for the getLimit function.

**Comment by wachterjohannes** [comment]
@danrot fixed

**Comment by wachterjohannes** [comment]
@danrot fixed and rebase

**Comment by danrot** [comment]
@wachterjohannes The build is always hanging at the same tests for PHP7...

**Comment by wachterjohannes** [comment]
@danrot this PR is now ready to review ... tests passes now without any changes

**Comment by danrot** [comment]
If the empty dropdown value from this screenshot is clicked, it is not written in the dropdown itself. This means it looks like the value `"` is still selected.

image

However, the export is correct.

**Comment by wachterjohannes** [comment]
@danrot the mentioned bug is fixed in the husky PR https://github.com/massiveart/husky/pull/617 this changes are also in this branch and can be tested here.
